src/speakers/extraction.py

```diff
@@ -329,3 +332,312 @@ def get_speakers_in_bold(doc):
+                    #modify_str takes the conversation, makes some changes and returns the words to w
+                    w=modify_str(l)
+                    #postion is the postion of the word identified in name 
+                    #name_index is the postion of the word identified in w
+                    count,count_max,position,name_index=find_count(name,w,count_max)
+                    if count/len(name.split())>=0.33 and count>=count_max:
+                        flag_2=0
+                        flag_3=0
+                        firm=''
+                        for c in range(name_index+len(name.split())-position,len(w)):
+                            if ((w[c][0].isupper() or w[c]=='individual') and flag_2==0 and w[c]!='Sir,' and w[c]!='I'):
+                                flag_2=1
+                                b=((w[c][-1]=='.' or w[c][-1]==','))
+                                while not b:
+                                    firm=firm+w[c]+' '
+                                    c=c+1
+                                    if c==len(w):
+                                        for letter in firm:
+                                            if letter.isnumeric():
+                                                firm=''
+                                        flag_3=1
+                                        break
+                                    b=((w[c][-1]=='.' or w[c][-1]==','))
+                                if flag_3==0:
+                                    firm=firm+w[c]
+                        firm=firm[0:len(firm)-1]                
+                        firm=check_if_firm(firm)
+                        if name[-1]==':':
+                            name=name[0:len(name)-1]
+                        final[name]=firm
+    return final
+
+
+def pass_3(names,conversation_from_beginning,converse,final):
+    for name in names:
+        count_max=0
+        name=remove_last_char_if_colon(name)
+        if name not in final.keys() or final[name]=='':
+            for i in range(3):
+                if name not in final.keys() or final[name]=='':
+                    count=0
+                    l=conversation_from_beginning[i]
+                    #modify_str takes the conversation, makes some changes and returns the words to w
+                    w=modify_str(l)
+                    #postion is the postion of the word identified in name 
+                    #name_index is the postion of the word identified in w
+                    count,count_max,position,name_index=find_count(name,w,count_max)
+                    if count/len(name.split())>=0.5 and count>=count_max:
+                        flag_2=0
+                        flag_3=0
+                        firm=''
+                        for c in range(name_index+len(name.split())-position,len(w)):
+                            if ((w[c][0].isupper() or w[c]=='individual') and flag_2==0 and w[c]!='Sir,' and w[c]!='Mr.'):
```

Comment:
If there's repeated logic b/w passes, let's reduce rewrite.

Reply:
There would be too many variables involved to write a meaningful function, also some conditions vary from pass to pass